README.md
# et_zookeeper cookbook

Wrapper cookbook for [SimpleFinance’s](https://github.com/SimpleFinance)
[zookeeper](https://github.com/SimpleFinance/chef-zookeeper) & [exhibitor](https://github.com/SimpleFinance/chef-exhibitor) cookbooks.

In addition to using these cookbooks to install Exhibitor and
Zookeeper, this cookbook sets up and enables [shared configuration using S3
for Exhibitor](https://github.com/Netflix/exhibitor/wiki/Shared-Configuration),
allowing Exhibitor to handle any additional configuration of the Zookeeper
ensemble.

# Requirements

* exhibitor

# Usage

Include the default recipe in a node’s run list.

If using with the [`apache_storm` cookbook](https://github.com/evertrue/apache_storm-cookbook), also ensure `role[zookeeper]` is included in the run list.

# Attributes

Attributes we override in the two base cookbooks:

* `node['zookeeper']['config']['dataDir']`
* `node['zookeeper']['config']['initLimit']`
* `node['zookeeper']['config']['syncLimit']`
* `node['zookeeper']['config']['autopurge.snapRetainCount']`
* `node['zookeeper']['config']['autopurge.purgeInterval']`
* `node['exhibitor']['version']`
* `node['exhibitor']['snapshot_dir']`
* `node['exhibitor']['transaction_dir']`
* `node['exhibitor']['log_index_dir']`
* `node['exhibitor']['cli']['configtype']`

Additional attributes provided by this cookbook:

* `node['exhibitor']['s3']['access-key-id']` – The IAM access key ID
* `node['exhibitor']['s3']['access-secret-key']` — The IAM secret access key
* `node['exhibitor']['cli']['s3region']` – AWS region used to communicate w/ S3
* `node['exhibitor']['cli']['s3config']` – S3 bucket & key name for the shared config file
* `node['et_exhibitor']['defaultconfig']['zoo_cfg_extra']` – Additional values used for the initial bootstrap; these are used to seed the S3 config file if it doesn’t already exist

# Recipes

## default

2. Sets additional Zookeeper config for Exhibitor to use
3. Includes `exhibitor::default`, which itself includes `zookeeper::install`, to install and set up Zookeeper and Exhibitor
4. Includes `exhibitor::service`, to set up and start the Exhibitor service via `runit`

# Author

Author:: EverTrue, Inc. (<devops@evertrue.com>)


CHANGELOG.md
et_zookeeper cookbook CHANGELOG
===============================
Changes for the et_zookeeper wrapper cookbook

## [v5.0.7] (2020-01-20)

* Removed filebeat setup

v5.0.6 (2017-02-10)
-------------------
* Make filebeat prospector an LWRP rather than an attribute

v5.0.5 (2017-02-10)
-------------------
* Add a filebeat prospector

v5.0.4 (2017-02-07)
-------------------
* Optimize cluster settings
    - `initLimit` to 120 seconds
    - `autopurge.snapRetainCount` to 3

v5.0.3 (2016-11-17)
-------------------
* Enable JMX

v5.0.2 (2016-11-10)
-------------------
* Pass java_opts to zookeeper_config resource
* Let application cookbook set the version

v5.0.1 (2016-11-09)
-------------------
* Bump initLimit to 100

v5.0.0 (2016-10-13)
-------------------
* Drop exhibitor cookbook (and associated attributes) and use ZK cookbook directly
* Bump ZK version: 3.4.9
* Use storage cookbook to manage ephemeral storage

v4.0.5 (2016-10-11)
-------------------
* Add an Ohai plugin to read myid into an attribute (allows for compatibility with what we hope will be v5)
* Modernize kitchen yaml

v4.0.4 (2016-10-06)
-------------------
* Re-enable reboot coordinator

v4.0.3 (2016-09-13)
-------------------
* Disable reboot coordinator

v4.0.2 (2016-09-07)
-------------------
* Test that datadog agent gets installed and configured
* Move datadog config to attributes file with default precedence so that it will actually run
* Switch to tcp-style test for datadog check
* Stop definining the cookbook attribute for datadog

v4.0.1 (2016-09-07)
-------------------
* Add Datadog attributes

v4.0.0 (2016-07-22)
-------------------
* Update `exhibitor` cookbook to `~> v2.0`
    - Also updates the `zookeeper` cookbook to `~> 5.0`
* Drop unnecessary dependencies
    - The `exhibitor` cookbook nicely wraps these, so simplifying the dep graph is a Good Thing™
* Clean up README
* Avoid using set/normal level attributes & drop unused/unnecessary attributes

v3.0.0 (2016-05-18)
-------------------
* Update `exhibitor` cookbook
    - Also updates the `zookeeper` cookbook, bringing in a bunch of fixes
* Fix up testing

v2.0.1 (2015-08-11)
-------------------
* Update exhibitor cookbook
    - Drop zookeeper as a dependency b/c exhibitor already depends on it
* Add `zookeeper-env.sh` attribute values to move `zookeeper.out` log & lower log noise
* Switch to running Test Kitchen using EC2
* Drop using real data bag items in favor of dummy items

v2.0.0 (2015-01-16)
-------------------
* Refactor using zookeeper ~> v2.0 & exhibitor ~> 0.4.0 cookbooks
* Clean up some configs

v1.2.1 (2014-11-06)
-------------------
* Update to installing Zookeeper 3.4.6
* Fix mirror to use for downloading ZK
* Use Supermarket for community cookbooks
* Fix Serverspec spec_helper

v1.2.0 (2014-05-05)
-------------------
* Update to installing Exhibitor 1.5.2
    * Fixes quite a few bugs, including one w/ the rolling restart
* Fix the zoo.cfg extra properties to match production
* Clean up unused config files
* Set up Test Kitchen properly
* Use Berkshelf 3
* Add Rakefile with handy testing:
    * Foodcritic
    * Rubocop
    * ChefSpec
    * Test Kitchen w/ ServerSpec
* Add ChefSpec tests
* Add ServerSpec tests


v1.1.6 (2014-01-15)
-------------
* Fix attributes used for S3 credentials

v1.1.5 (2013-01-14)
-------------
* Update zookeeper cookbook to ~> 1.4.9
* Adjust attributes to take advantage of fixes in [zookeeper v1.4.9](https://github.com/SimpleFinance/chef-zookeeper/compare/v1.4.8...v1.4.9) re: S3 Exhibitor config
* Remove unnecessary code related to setting up Exhibitor's S3 config

v1.1.4 (2013-01/07)
-------------
* Revert to Exhibitor 1.5.0 (1.5.1 doesn't stay up & running for some reason)
* Remove unnecessary attribute setting
* Clean up some code as per Rubocop

v1.1.3 (2013-01-06)
-------------
* Update to Exhibitor 1.5.1 (released 2013-10-01)

v1.1.2 (2013-12-10)
-------------
* Ensure Exhibitor auto-manages instances by default

v1.1.1 (2013-12-10)
-------------
* Bump version number to unstick servers that already got a bad v1.1.0 version

v1.1.0 (2013-12-10)
-------------
* Add S3 shared config for Exhibitor
* Don't redefine `exhibitor` service in order to enable it
* Clean up Ruby formatting as per Rubocop

v1.0.0 (2013-12-06)
-------------
* Bump version number to reflect production usage

v0.1.2 (2013-12-02)
-------------
* Clean up Gemfile
* Ensure using latest Berkshelf, Test Kitchen, Kitchen Vagrant gems

v0.1.1 (2013-11-11)
-------------
* Move data from /tmp to /mnt
* Translate zoo_cfg_extra from hash to string
* Set opts: autopurge.snapRetainCount=20, autopurge.purgeInterval=1

v0.1.0 (2013-11-02)
-------------
* First rev!


spec/recipes/default_spec.rb
require 'spec_helper'

describe 'et_zookeeper::default' do
  let(:chef_run) do
    ChefSpec::Runner.new(platform: 'ubuntu', version: '12.04').converge(described_recipe)
  end

  before do
    ChefSpec::Server.create_client('et_zookeeper_spec', admin: true)
    ChefSpec::Server.create_environment('_default', description: 'Default env')
    Chef::EncryptedDataBagItem.stub(:load).with('secrets', 'aws_credentials')
      .and_return(
        'ZookeeperS3' => {
          'access_key_id' => 'AAAAAAAAAAAAAAAAAAAA',
          'secret_access_key' => 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'
        }
      )
  end

  %w(
    apt::default
    build-essential::default
    zookeeper::default
  ).each do |recipe|
    it "includes #{recipe}" do
      expect(chef_run).to include_recipe recipe
    end
  end

  it 'creates a link to exhibitor’s upstart job' do
    link = chef_run.link('/etc/init.d/exhibitor')
    expect(link).to link_to('/lib/init/upstart-job')
  end

  it 'enables and starts exhibitor' do
    expect(chef_run).to enable_service('exhibitor')
    expect(chef_run).to start_service('exhibitor')
  end
end


spec/unit/recipes/default_spec.rb
require 'spec_helper'

describe 'et_zookeeper::default' do
  let(:chef_run) do
    ChefSpec::ServerRunner.new do |_node, server|
      server.create_client('et_zookeeper_spec', admin: true)
      server.create_environment('_default', description: 'Default env')
      server.create_data_bag(
        'secrets',
        'aws_credentials' => {
          'ZookeeperS3' => {
            'access_key_id' => 'AAAAAAAAAAAAAAAAAAAA',
            'secret_access_key' => 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'
          }
        }
      )
    end.converge(described_recipe)
  end

  %w(
    exhibitor::default
    exhibitor::service
  ).each do |recipe|
    it "includes #{recipe}" do
      expect(chef_run).to include_recipe recipe
    end
  end
end


test/integration/cookbooks/myid_checker/metadata.rb
name 'myid_checker'
version '1.0.1'


recipes/default.rb
#
# Cookbook Name:: et_zookeeper
# Recipe:: default
#
# Copyright (C) 2016 EverTrue, Inc.
#
# All rights reserved - Do Not Redistribute
#
include_recipe 'storage'

node.default['et_zookeeper']['home_dir'] =
  if node['storage']['ephemeral_mounts']
    "#{node['storage']['ephemeral_mounts'].first}/zookeeper"
  else
    '/mnt/zookeeper'
  end

node.default['et_zookeeper']['data_dir'] = "#{node['et_zookeeper']['home_dir']}/data"
node.default['et_zookeeper']['log_dir']  = "#{node['et_zookeeper']['home_dir']}/logs"

# Generate an ID from the node IP address
myid = EtZookeeper::Helpers.name2id node.name

# Find other nodes in the cluster
zk_nodes = search(
  :node,
  "chef_environment:#{node.chef_environment} AND " \
    "et_zookeeper_cluster_name:#{node['et_zookeeper']['cluster_name']}",
  filter_result: {
    'name' => %w(name),
    'myid' => %w(et_zookeeper myid),
    'ipaddress' => %w(ipaddress)
  }
)

# Bail if my ID is not unique in the cluster
zk_nodes.each do |zk_node|
  next unless zk_node['name'] != node.name && zk_node['myid'] == myid
  raise "myid #{myid} is already in use by #{zk_node['name']}"
end

node.default['et_zookeeper']['myid'] = myid

zookeeper node['zookeeper']['version'] do
  user_home node['et_zookeeper']['home_dir']
  data_dir node['et_zookeeper']['data_dir']
end

file "#{node['et_zookeeper']['data_dir']}/myid" do
  content myid.to_s
  owner   'zookeeper'
  group   'zookeeper'
end

config_hash = {
  'dataDir' => node['et_zookeeper']['data_dir'],
  'clientPort' => 2181,
  'initLimit' => 120,
  'syncLimit' => 17,
  'autopurge.snapRetainCount' => 3,
  'autopurge.purgeInterval' => 1
}.merge(
  zk_nodes.each_with_object({}) do |zk_node, m|
    m["server.#{zk_node['myid']}"] = "#{zk_node['name']}:2888:3888"
  end
).merge("server.#{myid}" => "#{node.name}:2888:3888") # Ensure the local node is always included

Chef::Log.debug("Zookeeper Config Hash: #{config_hash.inspect}")

zookeeper_config 'zoo.cfg' do
  config config_hash
  env_vars(
    'ZOO_LOG4J_PROP' => 'INFO,ROLLINGFILE',
    'JMXPORT' => node['et_zookeeper']['jmx_port'],
    'JMXSSL' => 'true',
    'JMXDISABLE' => !node['et_zookeeper']['jmx_enabled'] # lol double negatives. Thanks, Apache!
  )
  java_opts node['zookeeper']['java_opts']
  log_dir node['et_zookeeper']['log_dir']
end

zookeeper_service 'zookeeper'

ohai_plugin 'et_zookeeper'
ohai_plugin 'myid'

filebeat_prospector 'zookeeper' do
  paths ["#{node['et_zookeeper']['home_dir']}/logs/zookeeper.log"]
  document_type 'zookeeper'
  input_type 'log'
end
